src/structure/model.jl

```diff
@@ -0,0 +1,1089 @@
+    default::Char
+end
+
+
+"""
+A residue (amino acid) or other molecule represented in a PDB file - either a
+`Residue` or a `DisorderedResidue`.
+"""
+abstract AbstractResidue <: StructuralElement
+
+"A residue (amino acid) or other molecule record from a PDB file."
+immutable Residue <: AbstractResidue
+    name::ASCIIString
+    chain_id::Char
+    number::Int
+    ins_code::Char
+    het_res::Bool # Does the residue consist of hetatoms?
+    atom_list::Vector{ASCIIString}
+    atoms::Dict{ASCIIString, AbstractAtom}
+end
+
+# Constructor without atoms
+Residue(name::ASCIIString, chain_id::Char, number::Int, ins_code::Char, het_res::Bool) = Residue(
+    name, chain_id, number, ins_code, het_res, [], Dict())
+# Constructor from a list of atoms
+Residue{T <: AbstractAtom}(atoms::Vector{T}) = Residue(
+        resname(atoms[1]),
+        chainid(atoms[1]),
+        resnumber(atoms[1]),
+        inscode(atoms[1]),
+        ishetero(atoms[1]),
+        map(atomname, sort(atoms)),
+        [atomname(atom) => atom for atom in atoms]
+    )
+
+"A container to hold different versions of the same residue (point mutations)."
+immutable DisorderedResidue <: AbstractResidue
+    names::Dict{ASCIIString, Residue}
+    default::ASCIIString
+end
+
+
+"A chain represented in a PDB file."
+immutable Chain <: StructuralElement
+    id::Char
+    res_list::Vector{ASCIIString}
+    residues::Dict{ASCIIString, AbstractResidue}
+end
+
+Chain(id::Char) = Chain(id, [], Dict())
+
+
+"A conformation of a protein represented in a PDB file."
+immutable Model <: StructuralElement
+    number::Int
+    chains::Dict{Char, Chain}
+end
+
+Model(number::Int) = Model(number, Dict())
+Model() = Model(1)
+
+
+"A container for multiple models from a PDB file."
+immutable ProteinStructure <: StructuralElement
+    name::ASCIIString
+    models::Dict{Int, Model}
+end
+
+ProteinStructure(name::ASCIIString) = ProteinStructure(name, Dict())
+ProteinStructure() = ProteinStructure("")
+
+
+"A `StructuralElement` or `Vector` of `StructuralElement`s."
+typealias StructuralElementOrList Union{
+        StructuralElement,
+        Vector{AbstractAtom},
+        Vector{Atom},
+        Vector{DisorderedAtom},
+        Vector{AbstractResidue},
+        Vector{Residue},
+        Vector{DisorderedResidue},
+        Vector{Chain},
+        Vector{Model}
+    }
+
+
+# Allow accessing sub elements contained in an element like a dictionary
+# e.g. allows you to do res[atom_name] rather than res.atoms[atom_name]
+# setindex! should be used with caution as it can lead to inconsistencies
+# e.g. adding an atom to a residue atom dict does not update the atom list
+
+# Accessing a DisorderedAtom with a character returns the Atom with that alt
+#   loc ID
+Base.getindex(disordered_atom::DisorderedAtom, alt_loc_id::Char) = disordered_atom.alt_loc_ids[alt_loc_id]
+function Base.setindex!(disordered_atom::DisorderedAtom, atom::Atom, alt_loc_id::Char)
+    disordered_atom.alt_loc_ids[alt_loc_id] = atom
+end
```

Comment:
It would be better to return the modified object in `setindex!`, which is a convention in Julia.

```julia
function Base.setindex!(disordered_atom::DisorderedAtom, atom::Atom, alt_loc_id::Char)
    disordered_atom.alt_loc_ids[alt_loc_id] = atom
    return disordered_atom
end